### PR TITLE
Mec172x pinctrl

### PR DESCRIPTION
**Pull request by jvasanth1**
Implemented Pin Control for MEC172x.

This is resubmit; unable to submit new changes in previous PR #41888

### CODE REVIEW

File: dts/bindings/pinctrl/microchip,xec-pinctrl.yaml [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    The node has the 'pinctrl' node label set in your SoC's devicetree,
+    so you can modify it like this:
+
+      &pinctrl {
+              /* your modifications go here */
+      };
+
+    All device pin configurations should be placed in child nodes of the
+    'pinctrl' node, as shown in the spi0 example below:
+
```
Comment (jvasanth1) [inline dts/bindings/pinctrl/microchip,xec-pinctrl.yaml:21]: Updated `as shown in the example` to `as shown in the spi0 example below`

File: dts/bindings/pinctrl/microchip,xec-pinctrl.yaml [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    Driver strength and slew rate hardware defaults vary by SoC and pin.
+
+    An example for MEC172x family, include the chip level pinctrl
+    DTSI file in the board level DTS:
+
+      #include <microchip/mec172x/mec172xnsz-pinctrl.dtsi>
+
+    We want to use the shared SPI port of the MEC172x QMSPI controller
+    and want the chip select 0 to be open-drain.
+
```
Comment (jvasanth1) [inline dts/bindings/pinctrl/microchip,xec-pinctrl.yaml:47]: Fixed indentation.

File: dts/bindings/pinctrl/microchip,xec-pinctrl.yaml [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+          default: "low-speed"
+          enum:
+          - "low-speed"
+          - "high-speed"
+          description: |
+            Pin speed. The default value of slew-rate is the SoC power-on-reset
+            value. Please refer to the data sheet as a small number of pins
+            may have a different default and some pins do not implement
+            slew rate adjustment.
+
```
Comment (jvasanth1) [inline dts/bindings/pinctrl/microchip,xec-pinctrl.yaml:107]: Justified default value as `The default value of the slew-rate is the SoC power-on-reset value.`

File: boards/arm/mec172xevb_assy6906/mec172xevb_assy6906_defconfig [hunk 4]
```diff
@@ -12,6 +12,7 @@ CONFIG_RTOS_TIMER=y
 CONFIG_CLOCK_CONTROL=y
 CONFIG_GPIO=y
 CONFIG_PINMUX=y
+CONFIG_PINCTRL=y
```
Comment (gmarull) [inline boards/arm/mec172xevb_assy6906/mec172xevb_assy6906_defconfig:14]: looks like this change should not be in first commit

File: drivers/pinctrl/pinctrl_mchp_xec.c [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+	}
+
+	if (conf & BIT(MCHP_XEC_PIN_LOW_POWER_POS)) {
+		flags = GPIO_DISCONNECTED;
+	}
+
+	ret = gpio_pin_configure(gpio_ports[port], pin, flags);
+	if (ret) {
+		return ret;
+	}
```
Comment (gmarull) [inline drivers/pinctrl/pinctrl_mchp_xec.c:133]: please don't mix pinctrl/GPIO. pinctrl driver should be self-contained, so the logic to configure should be in this driver.

File: dts/bindings/pinctrl/microchip,xec-pinctrl.yaml (hunk 1 above)
Comment (gmarull) [reply dts/bindings/pinctrl/microchip,xec-pinctrl.yaml:21]: the sentence says "example below:", but there's no example just below. I guess it should say "in the spi0 example shown at the end."

File: soc/arm/microchip_mec/common/soc_i2c.c [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <sys/sys_io.h>
+#include <sys/__assert.h>
+#include <soc.h>
+#include "soc_i2c.h"
+
+#define MCHP_XEC_GPIO_REG_BASE	((struct gpio_regs *)0x40081000u)
```
Comment (gmarull) [inline soc/arm/microchip_mec/common/soc_i2c.c:13]: this information is already in DT, why is this needed? You can use DT API to obtain this one.

File: drivers/spi/spi_xec_qmspi_ldma.c [hunk 7]
```diff
@@ -1232,7 +1085,7 @@ static int qmspi_xec_init(const struct device *dev)
 	const struct spi_qmspi_config *cfg = dev->config;
 	struct spi_qmspi_data *qdata = dev->data;
 	struct qmspi_regs *regs = cfg->regs;
-
+	int ret = 0;
```
Comment (gmarull) [inline drivers/spi/spi_xec_qmspi_ldma.c:1090]: nit: `int ret;`

**Review by gmarull — CHANGES_REQUESTED** [review]

File: boards/arm/mec172xevb_assy6906/mec172xevb_assy6906_defconfig (hunk 4 above)
Comment (jvasanth1) [reply boards/arm/mec172xevb_assy6906/mec172xevb_assy6906_defconfig:14]: Only Pin control driver gets compiled on enabling this in the first commit. Ensured all commits are compiling before adding new commits.

File: drivers/pinctrl/pinctrl_mchp_xec.c (hunk 5 above)
Comment (jvasanth1) [reply drivers/pinctrl/pinctrl_mchp_xec.c:133]: we are making a API call to gpio driver; the function gpio_xec_configure is quite big, don't want to duplicate the code in pinctrl.

Comment (gmarull) [reply drivers/pinctrl/pinctrl_mchp_xec.c:133]: it's 87 LoC, whereas pinctrl "wrapper" is 61. I'd qualify as "quite big", so please make pinctrl driver independent.

**Review by jackrosenthal — DISMISSED** [review]
LGTM once other comments are resolved

File: drivers/spi/spi_xec_qmspi_ldma.c (hunk 7 above)
Comment (jvasanth1) [reply drivers/spi/spi_xec_qmspi_ldma.c:1090]: will update to 
`int ret;`

File: dts/bindings/pinctrl/microchip,xec-pinctrl.yaml (hunk 1 above)
Comment (jvasanth1) [reply dts/bindings/pinctrl/microchip,xec-pinctrl.yaml:21]: will update to `as in the spi0 example shown at the end.`

File: drivers/spi/spi_xec_qmspi_ldma.c (hunk 7 above)
Comment (jvasanth1) [reply drivers/spi/spi_xec_qmspi_ldma.c]: updated

File: soc/arm/microchip_mec/common/soc_i2c.c (hunk 6 above)
Comment (jvasanth1) [reply soc/arm/microchip_mec/common/soc_i2c.c]: using DT API

File: dts/bindings/pinctrl/microchip,xec-pinctrl.yaml (hunk 1 above)
Comment (jvasanth1) [reply dts/bindings/pinctrl/microchip,xec-pinctrl.yaml:21]: updated accordingly

File: drivers/pinctrl/pinctrl_mchp_xec.c (hunk 5 above)
Comment (jvasanth1) [reply drivers/pinctrl/pinctrl_mchp_xec.c:133]: @gmarull 
Could we take up pinctrl driver independent of gpio driver as a new PR? 
We would like to get all other pinctrl changes approved and merged first.

Comment (gmarull) [reply drivers/pinctrl/pinctrl_mchp_xec.c:133]: I see no reasons to keep a GPIO based pinctrl driver, so it's a -1 from my side. I already explained the reasons in the previous PR.

Comment (jvasanth1) [reply drivers/pinctrl/pinctrl_mchp_xec.c]: updated pinctrl to be independent of gpio driver; please review and approve.

**Review by gmarull — DISMISSED** [review]
LGTM, thanks for the continued work. @mbolivar-nordic @carlescufi compliance check failure is expected (pinctrl_soc_pin_t typedef)

**Comment by jvasanth1** [comment]
Thanks Gerard!

**Review by albertofloyd — APPROVED** [review]

**Review by jackrosenthal — DISMISSED** [review]

**Comment by albertofloyd** [comment]
LGTM. @jvasanth1 
I noticed that same KSCAN driver will be used for MEC152xx and MEC172x, has this been re-validated in MEC152xx?
This change breaks compilation when samples/helloworld is compiled for MEC152x EVB board with CONFIG_KSCAN=y
Will raise an issue, this needs to be address

File: boards/arm/mec172xevb_assy6906/mec172xevb_assy6906.dts [hunk 8]
```diff
@@ -130,25 +145,92 @@
...
+		      &shd_io0_gpio223
+		      &shd_io1_gpio224
+		      &shd_io2_gpio227
+		      &shd_io3_gpio016 >;
+	pinctrl-names = "default";
 };
 
 &kscan0 {
 	status = "okay";
+
```
Comment (albertofloyd) [inline boards/arm/mec172xevb_assy6906/mec172xevb_assy6906.dts:212]: @jvasanth1 since same driver will be used for MEC152x and MEC172x.
Driver should compile cleanly for both mec152x boards, that's not the case

**Review by albertofloyd — CHANGES_REQUESTED** [review]
Need to have separate KSCAN driver for MEC172x and MEC152x or fix compilation issue for MEC152x

File: boards/arm/mec172xevb_assy6906/mec172xevb_assy6906.dts (hunk 8 above)
Comment (albertofloyd) [reply boards/arm/mec172xevb_assy6906/mec172xevb_assy6906.dts:212]: @jvasanth1 will issue separately, need to understand why CI is not catching the issue

**Review by albertofloyd — DISMISSED** [review]

**Comment by enjiamai** [comment]
Hi @jvasanth1 , could you please rebase it? thank you!

**Comment by jvasanth1** [comment]
Rebased as per request; please help to re-approve.

**Comment by enjiamai** [comment]
Thanks! @jvasanth1 , could you please also fix the compliance check?  Thank you!
```
Run if [[ ! -s "compliance.xml" ]]; then
Error: See https://docs.zephyrproject.org/latest/contribute/#coding-style for more details.

-:3006: WARNING:NEW_TYPEDEFS: do not add new typedefs
#3006: FILE: soc/arm/microchip_mec/common/pinctrl_soc.h:[29](https://github.com/zephyrproject-rtos/zephyr/runs/5424671126?check_suite_focus=true#step:9:29):
+typedef struct {

- total: 0 errors, 1 warnings, 2969 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

Your patch has style problems, please review.
```

**Comment by jvasanth1** [comment]
The compliance check is not an issue. See @gmarull's comment [here](https://github.com/zephyrproject-rtos/zephyr/pull/43282#pullrequestreview-898182481).

**Review by gmarull — APPROVED** [review]

**Review by albertofloyd — APPROVED** [review]
